CenterForOpenScience/ember-osf-web
# [EMB-573] Fix VMF contention

## Purpose

Prevent multiple ValidatedModelForms on the same page using the same model from interfering with each other.

## Summary of Changes

1. Add `ember-changesets` addon
2. Modify `validated-model-form` to use `ember-changesets`
3. Fix the methods that failed because we are returning a proxy rather than a model
4. Update the handbook
5. Add a test specifically for EMB-573
6. Update connected-email model to use changeset. Note: this appears to be not tested and isn't miraged, so this might not work. I figured we could handle that in a separate PR. Also removed the debounce on validation because a) premature optimization; b) doesn't work exactly the way you think; and most importantly c) does not work at *all* with changesets. Ends up debouncing the user input rather than the validation.
7. `validated-input/base-component.ts` and `x-input-wrapper` had to be modified to work with both models and changesets. If we ever ditch non VMF forms, we can remove the model-based code.

## Side Effects

- If there's an instance of `validated-model-form` that uses an untested method that's relying on dot access to a model's properties, it will fail when it runs.
- isPageDirty tracking that checks the state of the model will fail when using changesets, but the Handbook shows how to do it properly.
- Forms using VMF will no longer live-update the model, so if you are editing something that is also rendered on the screen, the rendered version will not change until a successful submission of the form. (This is technically the removal of a side effect)

## QA Notes

This affects only existing ValdatedModelForms. That means that…

- developer apps
- tokens
- Registries comment card
- Collections project metadata

…may need regression testing when this is merged. Also, find any set of checkboxes and make sure it does what it's supposed to.

## Ticket


https://openscience.atlassian.net/browse/EMB-573

# Reviewer Checklist
